File: pew_customizations/crm/customer.py
import frappe
from frappe import _
from frappe.utils import cint

from pew_customizations.crm.config import ACTIVE_STATUSES, COLD_STAGE, WON_STAGE
from pew_customizations.crm.opportunity import get_pipeline_counts


def onload(doc, method=None):
	"""Live Opportunity counts for the "Tender Pipeline" section of the Customer dashboard."""
	if doc.is_new() or not frappe.has_permission("Opportunity", "read"):
		return
	counts = get_pipeline_counts("Customer", doc.name)
	doc.set_onload(
		"pew_opportunity_stats",
		{
			"counts": {key: cint(value) for key, value in counts.items()},
			"stages": {"cold": COLD_STAGE, "won": WON_STAGE},
			"active_statuses": list(ACTIVE_STATUSES),
		},
	)


def repoint_source_opportunity(doc, method=None):
	"""Customer made with Opportunity > Create > Customer: move that Opportunity onto the new Customer
	so it passes the Inquiry / Tender gate and is counted on the Customer dashboard."""
	if not doc.opportunity_name or not frappe.db.exists("Opportunity", doc.opportunity_name):
		return
	opportunity_from = frappe.db.get_value("Opportunity", doc.opportunity_name, "opportunity_from")
	if opportunity_from == "Customer":
		return
	frappe.db.set_value(
		"Opportunity",
		doc.opportunity_name,
		{"opportunity_from": "Customer", "party_name": doc.name, "customer_name": doc.customer_name},
	)
	frappe.get_doc("Opportunity", doc.opportunity_name).add_comment(
		"Info", _("Organization changed from {0} to Customer {1}.").format(opportunity_from, doc.name)
	)

File: pew_customizations/public/js/customer.js
// "Tender Pipeline" counters on the Customer dashboard. Counts are computed live on load
// (pew_customizations/crm/customer.py); each counter opens the matching Opportunity list.

frappe.ui.form.on("Customer", {
	refresh(frm) {
		const stats = frm.doc.__onload && frm.doc.__onload.pew_opportunity_stats;
		if (frm.is_new() || !stats) return;

		const { counts, stages, active_statuses } = stats;
		const base = { opportunity_from: "Customer", party_name: frm.doc.name };
		const tiles = [
			{ key: "total", label: __("Total Inquiries"), color: "blue", filters: {} },
			{ key: "active", label: __("Active Bids"), color: "orange", filters: {
				status: ["in", active_statuses],
				sales_stage: ["not in", [stages.cold, stages.won]],
			} },
			{ key: "won", label: __("Won"), color: "green", filters: {
				sales_stage: stages.won,
				status: ["!=", "Lost"],
			} },
			{ key: "lost", label: __("Lost"), color: "red", filters: { status: "Lost" } },
			{ key: "cold", label: __("Cold"), color: "gray", filters: {
				sales_stage: stages.cold,
				status: ["!=", "Lost"],
			} },
		];

		const html = `<div class="row">${tiles
			.map(
				(t) => `<div class="col-sm-2 col-xs-6">
					<a class="pew-pipeline-tile" data-key="${t.key}" href="#">
						<div class="h4 mb-1 text-${t.color}">${cint(counts[t.key])}</div>
						<div class="small text-muted">${t.label}</div>
					</a>
				</div>`
			)
			.join("")}</div>`;

		const $section = frm.dashboard.add_section(html, __("Tender Pipeline"));
		$section.on("click", ".pew-pipeline-tile", (e) => {
			e.preventDefault();
			const tile = tiles.find((t) => t.key === $(e.currentTarget).data("key"));
			frappe.set_route("List", "Opportunity", { ...base, ...tile.filters });
		});
		frm.dashboard.show();
	},
});
